### PR TITLE
chapter_8

### PR DESCRIPTION
**Pull request by chuan97sun**

### CODE REVIEW

**Comment by chuan97sun** [comment]
ready

File: app/assets/stylesheets/sessions.scss [hunk 1]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Sessions controller here.
```
Comment (manhnt-0960) [inline app/assets/stylesheets/sessions.scss]: delete this empty file

File: app/controllers/sessions_controller.rb [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+  def new; end
+
+  def create
+    user = User.find_by(email: params[:session][:email].downcase)
+    if user&.authenticate(params[:session][:password])
+      log_in user
+      redirect_to user
+    else
+      flash[:danger] = "Invalid email/password combination" # Not quite right!
+      render 'new'
```
Comment (manhnt-0960) [inline app/controllers/sessions_controller.rb]:
```suggestion
      render :new
```

File: app/helpers/sessions_helper.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
+module SessionsHelper
+
```
Comment (manhnt-0960) [inline app/helpers/sessions_helper.rb]: remove this empty line

File: app/helpers/sessions_helper.rb [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+  end
+
+  def current_user
+    if session[:user_id]
+      @current_user ||= User.find_by(id: session[:user_id])
+    end
+  end
+
+  def logged_in?
+    !current_user.nil?
```
Comment (manhnt-0960) [inline app/helpers/sessions_helper.rb]:
```suggestion
    current_user.present?
```

File: app/helpers/sessions_helper.rb [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+  end
+
+  def log_out
+    session.delete(:user_id)
+    @current_user = nil
+ end
+end
```
Comment (manhnt-0960) [inline app/helpers/sessions_helper.rb]: please add newline at end of the file

File: app/models/user.rb [hunk 6]
```diff
@@ -23,4 +23,9 @@ def downcase_email
     email.downcase!
   end
 
+  def User.digest(string)
```
Comment (manhnt-0960) [inline app/models/user.rb]: use syntax `class << sefl` when define class method

File: app/views/layouts/_header.html.erb [hunk 7]
```diff
@@ -1,12 +1,29 @@
...
+            </ul>
+          </li>
+        <% else %>
+          <li><%= link_to "Log in", login_path %></li>
+        <% end %>
+      </ul>
+    </nav>
   </div>
-</header>
+</header>
```
Comment (manhnt-0960) [inline app/views/layouts/_header.html.erb]: add newline at end of file

File: app/views/sessions/new.html.erb [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+        <%= f.label :email %>
+        <%= f.email_field :email, class: 'form-control' %>
+        <%= f.label :password %>
+        <%= f.password_field :password, class: 'form-control' %>
+        <%= f.submit "Log in", class: "btn btn-primary" %>
+
+    <% end %>
+    <p>New user? <%= link_to "Sign up now!", signup_path %></p>
+  </div>
+</div
```
Comment (manhnt-0960) [inline app/views/sessions/new.html.erb]: semilar

File: config/routes.rb [hunk 9]
```diff
@@ -1,6 +1,10 @@
 Rails.application.routes.draw do	
+  get 'sessions/new'
```
Comment (manhnt-0960) [inline config/routes.rb]: remove this line because it's semilar line 7

File: app/views/sessions/new.html.erb (hunk 8 above)
Comment (chuan97sun) [reply app/views/sessions/new.html.erb]: Sorry for missing the newline,was spending time to check indent

**Comment by chuan97sun** [comment]
ready

File: app/controllers/application_controller.rb [hunk 10]
```diff
@@ -1,2 +1,4 @@
 class ApplicationController < ActionController::Base
+  include SessionsHelper
 end
+
```
Comment (saobv-2223) [inline app/controllers/application_controller.rb]: remove this line

File: app/controllers/sessions_controller.rb [hunk 11]
```diff
@@ -0,0 +1,19 @@
+class SessionsController < ApplicationController
+  def new; end
+
+  def create
+    user = User.find_by(email: params[:session][:email].downcase)
```
Comment (saobv-2223) [inline app/controllers/sessions_controller.rb]: remove ()

File: app/helpers/sessions_helper.rb [hunk 12]
```diff
@@ -0,0 +1,20 @@
+module SessionsHelper
+  def log_in(user)
```
Comment (saobv-2223) [inline app/helpers/sessions_helper.rb]: remove ()

File: app/helpers/sessions_helper.rb [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+      @current_user ||= User.find_by(id: session[:user_id])
+    end
+  end
+
+  def logged_in?
+    current_user.present?
+  end
+
+  def log_out
+    session.delete(:user_id)
```
Comment (saobv-2223) [inline app/helpers/sessions_helper.rb]: remove ()

File: app/javascript/packs/application.js [hunk 14]
```diff
@@ -7,6 +7,8 @@ require("@rails/ujs").start()
 require("turbolinks").start()
 require("@rails/activestorage").start()
 require("channels")
+require("jquery")
+import "bootstrap"
```
Comment (saobv-2223) [inline app/javascript/packs/application.js]: using '

File: app/models/user.rb [hunk 15]
```diff
@@ -23,4 +23,11 @@ def downcase_email
     email.downcase!
   end
 
+  class << self
+    def User.digest(string) 
+      cost = ActiveModel::SecurePassword.min_cost ? BCrypt::Engine::MIN_COST :
+        BCrypt::Engine.cost
+        BCrypt::Password.create(string, cost: cost)
```
Comment (saobv-2223) [inline app/models/user.rb]: remove ()

File: app/views/layouts/_header.html.erb [hunk 16]
```diff
@@ -1,12 +1,29 @@
...
+              <li><%= link_to "Profile", current_user %></li>
+              <li><%= link_to "Settings", '#' %></li>
+              <li class="divider"></li>
+              <li>
+                <%= link_to "Log out", logout_path, method: :delete %>
+              </li>
+            </ul>
+          </li>
+        <% else %>
+          <li><%= link_to "Log in", login_path %></li>
```
Comment (saobv-2223) [inline app/views/layouts/_header.html.erb]: using "

File: app/views/sessions/new.html.erb [hunk 17]
```diff
@@ -0,0 +1,17 @@
+<% provide :title, "Log in" %>
+<h1>Log in</h1>
+
+  <div class="row">
+    <div class="col-md-6 col-md-offset-3">
+      <%= form_for(:session, url: login_path) do |f| %>
```
Comment (saobv-2223) [inline app/views/sessions/new.html.erb]: remov e()

File: app/views/sessions/new.html.erb [hunk 18]
```diff
@@ -0,0 +1,17 @@
...
+
+        <%= f.label :email %>
+        <%= f.email_field :email, class: 'form-control' %>
+        <%= f.label :password %>
+        <%= f.password_field :password, class: 'form-control' %>
+        <%= f.submit "Log in", class: "btn btn-primary" %>
```
Comment (saobv-2223) [inline app/views/sessions/new.html.erb]: using "

**Comment by chuan97sun** [comment]
ready

File: app/controllers/sessions_controller.rb [hunk 19]
```diff
@@ -0,0 +1,19 @@
+class SessionsController < ApplicationController
+  def new; end
+
+  def create
+    user = User.find_by email: params[:session][:email].downcase
+    if user&.authenticate(params[:session][:password])
+      log_in user
+      redirect_to user
+    else
+      flash[:danger] = "Invalid email/password combination"
```
Comment (chungpv-1008) [inline app/controllers/sessions_controller.rb]:
```suggestion
      flash.now[:danger] = "Invalid email/password combination" 
```

File: app/helpers/sessions_helper.rb [hunk 20]
```diff
@@ -0,0 +1,20 @@
+module SessionsHelper
+  def log_in user
+    session[:user_id] = user.id
+  end
+
+  def current_user
+    if session[:user_id]
+      @current_user ||= User.find_by(id: session[:user_id])
```
Comment (chungpv-1008) [inline app/helpers/sessions_helper.rb]:
```suggestion
      @current_user ||= User.find_by id: session[:user_id]
```

**Comment by chuan97sun** [comment]
ready

**Comment by ngocvt-0484** [comment]
merged

**Pull request merged**